highfidelity/hifi
# findRayIntersection/findParabolaIntersection performance improvements

Improves picking performance in certain cases:
- Octree cells are recursed in order of bounding box distance, instead of always in the same order, so farther cells don't need to be checked
- Model TriangleSets are checked in order of bounding box distance, instead of in a random order.
- TriangleSets store triangles in a simple k-d tree, instead of an octree, for a better distribution of triangles.  Cells are always split in half along the longest dimension.  When walking the tree, the two children are sorted by bounding box distance.
- Avatars are checked in order of capsule distance, instead of always checking all of them.
- Faster triangle intersection code (Möller-Trumbore)
- https://highfidelity.manuscript.com/f/cases/17116/implement-faster-ray-box-algorithm

Adds a menu option (Developer -> Picking -> Force Coarse Picking) to force Picks to always do coarse intersection, so we can see the price of precise picking against models.

Test plan:
- Run the https://github.com/highfidelity/hifi_tests/tree/master/tests/engine/interaction/pointer tests.
- In an HMD, use far grab and teleport and verify that they work as in master.
- Go to a completely empty domain.
- Make sure you only have default scripts running and open the expanded stats.  At the bottom of the first column, you should see "Intersection calls:" and then below that, "Rays".  It should say 6/4/0/0.
- Run [this](https://gist.githubusercontent.com/SamGondelman/4a2eb74273cf19c58d8ff933db0f8e7c/raw/1f40ece4364de73259fd3efee43c1645713870b5/PickPerformanceTest.js).  You should see lots of models appear in front of you.  (If necessary, open up the LOD tools (Developer -> Render -> LOD Tools), set it to manual, and move the slider all the way to the right, so the LOD stops oscillating.)
- You should see 7/4/0/0 in that row.
- In desktop, move to each corner of the cube of models and point towards the center, so the ray from your mouse would intersect the longest line of models.  You should see a blue sphere obstructing your view and another blue sphere at the intersection point, if there is one.  Watch the pickManager time in the first column and wait for it to settle.  Record it for each corner of the cube.

Compare to https://github.com/highfidelity/hifi/pull/13876/:
- In https://github.com/highfidelity/hifi/pull/13876/, running that script, the stats should say 8/4/0/0
- Repeat the last step from above with https://github.com/highfidelity/hifi/pull/13876/.  The times should be higher in that PR, usually by a factor of about 2-3.
